library/core/src/main/java/com/google/android/exoplayer2/upstream/DefaultHttpDataSource.java

```diff
@@ -377,8 +377,8 @@ public long open(DataSpec dataSpec) throws HttpDataSourceException {
         errorResponseBody =
             errorStream != null ? Util.toByteArray(errorStream) : Util.EMPTY_BYTE_ARRAY;
       } catch (IOException e) {
-        throw new HttpDataSourceException(
-            "Error reading non-2xx response body", e, dataSpec, HttpDataSourceException.TYPE_OPEN);
+        throw new InvalidResponseCodeException(
```

Comment:
Same comment as in `OkHttpDataSource`. I think it's better just to set `errorResponseBody` to Util.EMPTY_BYTE_ARRAY` and fall through.

Note that `CronetDataSource` should also be modified as part of this change. Probably by enclosing [this while loop](https://github.com/google/ExoPlayer/blob/1347d572ef9ce79aacd667cfffa7d7468c8408a4/extensions/cronet/src/main/java/com/google/android/exoplayer2/ext/cronet/CronetDataSource.java#L555) in a `try/catch` block and then in the `catch` block setting `responseBody` back to `Util.EMPTY_BYTE_ARRAY` before falling through.